toolchain/check/handle_binding_pattern.cpp

```diff
@@ -241,21 +255,29 @@ auto HandleParseNode(Context& context,
 }
 
 auto HandleParseNode(Context& context, Parse::AddrId node_id) -> bool {
-  auto self_param_id = context.node_stack().PopPattern();
-  if (auto self_param =
-          context.insts().TryGetAs<SemIR::AnyBindName>(self_param_id);
-      self_param &&
-      context.entity_names().Get(self_param->entity_name_id).name_id ==
-          SemIR::NameId::SelfValue) {
-    // TODO: The type of an `addr_pattern` should probably be the non-pointer
-    // type, because that's the type that the pattern matches.
-    context.AddInstAndPush<SemIR::AddrPattern>(
-        node_id, {.type_id = self_param->type_id, .inner_id = self_param_id});
+  auto param_pattern_id = context.node_stack().PopPattern();
+  if (SemIR::Function::GetParamPatternInfoFromPatternId(context.sem_ir(),
+                                                        param_pattern_id)
+          .GetNameId(context.sem_ir()) == SemIR::NameId::SelfValue) {
+    auto pointer_type = context.types().TryGetAs<SemIR::PointerType>(
+        context.insts().Get(param_pattern_id).type_id());
+    if (pointer_type) {
+      auto addr_pattern_id = context.AddPatternInst<SemIR::AddrPattern>(
+          node_id,
+          {.type_id = pointer_type->pointee_id, .inner_id = param_pattern_id});
+      context.node_stack().Push(node_id, addr_pattern_id);
+    } else {
+      CARBON_DIAGNOSTIC(
+          AddrOnNonPointerType, Error,
+          "`addr` can only be applied to a binding with a pointer type");
+      context.emitter().Emit(node_id, AddrOnNonPointerType);
+      context.node_stack().Push(node_id, param_pattern_id);
```

Comment:
Hm, I see. Yeah, I think I was wrong to think that we don't need types for patterns until we see the scrutinee. As you say, for parameters we really do need a type, and indeed we presumably want a rule for parameters that disallows any case where the type is not known -- or where matching against an argument of a different type would do something that's not equivalent to converting the argument to the parameter type and then matching it.

That means that we can support binding patterns, `var` patterns, tuple patterns, and struct patterns in function parameter lists, but not expression patterns, the designator form of alternative patterns, nor `auto` deduction. That sounds right to me, and has a lot of overlap with not wanting refutable patterns in function parameters (which is mostly the same but also rules out the fully-qualified form of alternative patterns but would allow `auto`).

OK. So with this new understanding, I think I agree, something like `BuiltinAutoType` makes sense.

Aside: I also wonder whether we'll want a distinct `ExprCategory` for patterns -- rather than treating them as simply `NotExpr`s with a type, classifying them as `ExprCategory::Pattern` might allow code to more easily reason about them. But let's defer that until we have an actual use for that information. Also, maybe modeling this separately from the `ExprCategory` is better, to avoid adding churn to the `switch`es on `ExprCategory`.

Reply:
OK, so is this resolved?